lib/execution_segment.go

```diff
@@ -430,199 +429,177 @@ func gcd(a, b int64) int64 {
+	sort.SliceStable(sortedNormalizedIndexes, func(i, j int) bool {
+		return sortedNormalizedIndexes[i].normNumerator > sortedNormalizedIndexes[j].normNumerator
 	})
 
-	if index < 0 || index >= len(et.sequence) || !et.sequence[index].Equal(segment) {
-		return -1
-	}
-	return index
-}
-
-// ScaleInt64 scales the provided value based on the ExecutionTuple
-func (et *ExecutionTuple) ScaleInt64(value int64) int64 {
-	if et.esIndex == -1 {
-		return 0
+	// This is the striping algorithm. Imagine you have a number of rational
+	// numbers which all add up to 1 (or less), and call them segments. If you
+	// want each to get proportional amount of anything, you need to give them
+	// their numerator count of elements for each denominator amount from the
+	// original elements. So, for 1/3, you give 1 element for each 3 elements.
+	// For 3/5 - 3 elements for each 5. If you have, for example, a sequence
+	// with elements with length 3/5 and 1/3, in order to know how to distribute
+	// it accurately, you need to get the LCD(lowest common denominitor). In
+	// this case, between 3 and 5, the LCD is 15. Then to transform the numbers
+	// to have the same, LCD equal, denominator. So 3/5 becomes 9/15 and 1/3
+	// becomes 5/15. So now for each 15 elements 9 need to go to the 3/5, and 5
+	// need to go to 1/3. This is what we did above in sortedNormalizedIndexes.
+	//
+	// We use the algorithm below to split elements between ExecutionSegments by
+	// using their length as the rational number. As we would like to get
+	// non-sequential elements, we try to get the maximum distance between them.
+	// That is the number of elements divided by the number of elements for any
+	// given segment, which concidently is the length of the segment reversed.
+	// The algorithm below does the following:
+	//  1. Goes through the elements from 0 to the lcd-1
+	//  2. For each of element, it goes through the segments and looks if the
+	//     amount of already taken elements by the given segment, multiplied by
+	//     that segment's length inverted, is equal to or less to the current
+	//     element index. If it is, give that element to that segment. If not,
+	//     continue with the next element.
+	// The code below specifically avoids using big.Rat, for performance
+	// reasons, which complicates the code somewhat. As additional note, the
+	// sorting of the segments from biggest to smallest helps with the fact that
+	// the biggest elements will need to take the most elements, and for them it
+	// will be the hardest to not get sequential elements.
+	prev := make([]int64, sequenceLength)
+	chosenCounts := make([]int64, sequenceLength)
+	saveIndex := func(iteration int64, index int, numerator int64) {
+		offsets[index] = append(offsets[index], iteration-prev[index])
+		prev[index] = iteration
+		if int64(len(offsets[index])) == numerator {
+			offsets[index] = append(offsets[index], offsets[index][0]+lcd-iteration)
+		}
```

Comment:
I like that this lambda is not external anymore, and the algorithm is in a single place now. :+1: